copier/copier.go

```diff
@@ -0,0 +1,1367 @@
+		n, err := io.Copy(f, tr)
+		if err != nil {
+			return n, errors.Wrapf(err, "copier: put: error writing file %q", path)
+		}
+		if err = f.Chmod(mode); err != nil {
+			return n, errors.Wrapf(err, "copier: put: error setting permissions on file %q", path)
+		}
+		return n, nil
+	}
+	targetDirectory, err := resolvePath(request.Root, request.Directory, nil)
+	if err != nil {
+		return errorResponse("copier: put: error resolving %q: %v", request.Directory, err)
+	}
+	info, err := os.Lstat(targetDirectory)
+	if err == nil {
+		if !info.IsDir() {
+			return errorResponse("copier: put: %s (%s): exists but is not a directory", request.Directory, targetDirectory)
+		}
+	} else {
+		if !os.IsNotExist(err) {
+			return errorResponse("copier: put: %s: %v", request.Directory, err)
+		}
+		if err := ensureDirectoryUnderRoot(request.Directory); err != nil {
+			return errorResponse("copier: put: %v", err)
+		}
+	}
+	cb := func() error {
+		type directoryAndTimes struct {
+			directory    string
+			atime, mtime time.Time
+		}
+		var directoriesAndTimes []directoryAndTimes
+		defer func() {
+			for i := range directoriesAndTimes {
+				directoryAndTimes := directoriesAndTimes[len(directoriesAndTimes)-i-1]
+				if err := lutimes(false, directoryAndTimes.directory, directoryAndTimes.atime, directoryAndTimes.mtime); err != nil {
+					logrus.Debugf("error setting access and modify timestamps on %q to %s and %s: %v", directoryAndTimes.directory, directoryAndTimes.atime, directoryAndTimes.mtime, err)
+				}
+			}
+		}()
+		tr := tar.NewReader(bulkReader)
+		hdr, err := tr.Next()
+		for err == nil {
+			// figure out who should own this new item
+			if idMappings != nil && !idMappings.Empty() {
+				containerPair := idtools.IDPair{UID: hdr.Uid, GID: hdr.Gid}
+				hostPair, err := idMappings.ToHost(containerPair)
+				if err != nil {
+					return errors.Wrapf(err, "error mapping container filesystem owner 0,0 to host filesystem owners")
+				}
+				hdr.Uid, hdr.Gid = hostPair.UID, hostPair.GID
+			}
+			if hdr.Typeflag == tar.TypeDir {
+				if request.PutOptions.ChownDirs != nil {
+					hdr.Uid, hdr.Gid = dirUID, dirGID
+				}
+			} else {
+				if request.PutOptions.ChownFiles != nil {
+					hdr.Uid, hdr.Gid = *fileUID, *fileGID
+				}
+			}
+			// make sure the parent directory exists
+			path := filepath.Join(targetDirectory, string(os.PathSeparator)+hdr.Name)
+			if err := ensureDirectoryUnderRoot(filepath.Dir(path)); err != nil {
+				return err
+			}
+			// figure out what the permissions should be
+			mode := os.FileMode(hdr.Mode) & os.ModePerm
+			if hdr.Typeflag == tar.TypeDir {
+				if request.PutOptions.ChmodDirs != nil {
+					hdr.Mode = int64(*request.PutOptions.ChmodDirs)
+				}
+			} else {
+				if request.PutOptions.ChmodFiles != nil {
+					hdr.Mode = int64(*request.PutOptions.ChmodFiles)
+				}
+			}
+			// create the new item
+			devMajor := uint32(hdr.Devmajor)
+			devMinor := uint32(hdr.Devminor)
+			written := hdr.Size
+			switch hdr.Typeflag {
```

Comment:
Nothing prevents `path` from pointing outside of `Root` AFAICS.

Reply:
Started prepending the path separator to `hdr.Name` and running the result through `filepath.Clean()`, which should first remove any internal ".." references, and then remove any initial ones, before joining the result with the top level directory.  I _think_ that's an improvement.